bin/gen_prediction_chunks.py

```diff
@@ -73,54 +91,94 @@ def write_chunks(data, pbar=None):
-    print(" done.", file = sys.stderr, flush=True)
-    print("Writing output files...", file = sys.stderr, flush=True)
-
-    # Write the necessary predictions into chunks of peptide lists
+    print("\nJoining protein ids with allele info...", flush=True)
+    # Prepare df with allele info for proteins for downstream merging against all peptides
+    proteins_allele_info = entities_proteins_occs\
+        .merge(microbiomes_entities_occs)\
+        .drop(columns="entity_id")\
+        .merge(conditions)\
+        .drop(columns="microbiome_id")\
+        .merge(condition_allele_map)\
+        .drop(columns="condition_id")\
+        .merge(alleles)\
+        .drop_duplicates()\
+        .set_index('protein_id')\
+        .sort_index()
+    # -> protein_id, allele_id, allele_name
+
+    print("\nInfo: proteins_allele_info", flush=True)
+    proteins_allele_info.info(verbose = False, memory_usage=print_mem)
+
     cur_chunk = 0
-    to_predict.groupby("allele_id").apply(lambda x : write_chunks(x))
+    requests = 0
+    keep = pd.DataFrame()
+    # Process peptides chunk-wise, write out into files with max_chunk_size or keep for next chunk if remaining requests, i.e. < max_chunk_size for one allele
+    with pd.read_csv(args.peptides, sep='\t', index_col="peptide_id", chunksize=args.proc_chunk_size) as reader:
+        for c, peptides in enumerate(reader):
+            print("\nChunk: ", c)
+            print("Info: peptides", flush=True)
+            peptides.info(verbose = False, memory_usage=print_mem)
+
+            # Identify which predictions have to be computed: join peptides with allele info
+            to_predict = peptides\
+                    .sort_index()\
+                    .join(protein_peptide_occs)\
+                    .reset_index(level='peptide_id')\
+                    .set_index('protein_id')\
+                    .sort_index()\
+                    .join(proteins_allele_info)\
+                    .drop_duplicates()\
+                    .reset_index(drop=True)
+            # -> index, peptide_id, peptide_sequence, allele_id, allele_name
+
+            # TODO peptides can be deleted?
+
+            print("Info: to_predict", flush=True)
+            to_predict.info(verbose = False, memory_usage=print_mem)
+
+            requests += len(to_predict)
+
+            # Concat remaining peptides from previous round with newly processed ones
+            # write the required predictions into chunks of peptide lists
+            keep = pd.concat([keep, to_predict], ignore_index=True)\
+                .groupby("allele_id", group_keys=False)\
+                .apply(lambda x : write_chunks(x))
+            # use group_keys=False to avoid generation of extra index with "allele_id"
+
+            print("Info: keep", flush=True)
+            keep.info(verbose = False, memory_usage=print_mem)
+
+        # Write out remaining peptides
+        keep.groupby("allele_id", group_keys=False).apply(lambda x : write_chunks(x, remainder=True))
```

Comment:
Doesn't have to happen in the `reader` scope
```suggestion
      # Write out remaining peptides
      keep.groupby("allele_id", group_keys=False).apply(lambda x : write_chunks(x, remainder=True))
```